### PR TITLE
140035 - Bess Configuration Service Frequency Expression / Cron Implementation

### PR DESCRIPTION
**Pull request by MastekRohitChettri**
Implement the logic to read Frequency expression i.e. CRON, so as to be useful for generating Bespoke ESS as per the frequency.

PBI consists changes of [PBI 140035](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/140035)

### CODE REVIEW

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs [hunk 1]
```diff
@@ -82,5 +92,81 @@ private bool IsValidJson(string json, string fileName)
...
+
+                ScheduleDetailEntity scheduleDetailEntity = new();
+                {
+                    scheduleDetailEntity.NextScheduleTime = nextOccurrence;
+                    scheduleDetailEntity.IsEnabled = bessConfig.IsEnabled;
+                }
+
+                return scheduleDetailEntity;
+            }
+            return existingScheduleDetail;
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs:172]:
```suggestion
            var existingScheduleDetail = azureTableStorageHelper.GetScheduleDetail(bessConfig.Name);

if (!string.IsNullOrEmpty(existingScheduleDetail.RowKey))
{
    return existingScheduleDetail;
}

azureTableStorageHelper.UpsertScheduleDetail(nextOccurrence, bessConfig, false);

ScheduleDetailEntity scheduleDetailEntity = new();
{
    scheduleDetailEntity.NextScheduleTime = nextOccurrence;
    scheduleDetailEntity.IsEnabled = bessConfig.IsEnabled;
}

return scheduleDetailEntity;
```

**Review by Nadeem14456 — CHANGES_REQUESTED** [review]
Looks good. Just a couple of suggestions from my end.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AzureTableStorageHelper.cs [hunk 2]
```diff
@@ -75,5 +77,27 @@ private TableClient GetTableClient(string tableName)
...
+            };
+
+            TableClient tableJobScheduleEntityClient = GetTableClient(BESS_SCHEDULE_DETAILS_TABLE_NAME);
+            tableJobScheduleEntityClient.UpsertEntity(scheduleDetailEntity);
+        }
+
+        public ScheduleDetailEntity GetScheduleDetail(string name)
+        {
+            TableClient tableJobScheduleEntityClient = GetTableClient(BESS_SCHEDULE_DETAILS_TABLE_NAME);
+            ScheduleDetailEntity scheduleDetailEntity = tableJobScheduleEntityClient.Query<ScheduleDetailEntity>().Where(i => i.IsEnabled.Equals(true) && i.RowKey.Equals(name)).FirstOrDefault();
```
[the remark this answers is not in the document]
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AzureTableStorageHelper.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs [hunk 3]
```diff
@@ -435,7 +435,15 @@ public enum EventIds
...
-        BessErrorOccurredWhileDownloadingConfigFromAzureStorage = 809107
+        BessErrorOccurredWhileDownloadingConfigFromAzureStorage = 809107,
+        /// <summary>
+        /// 809108 - Schedule config details matched with current time
+        /// </summary>
+        BessConfigFrequencyElapsed = 809108,
+        /// <summary>
+        /// 809109 - Exception occurred while processing bespoke details to queue
+        /// </summary>
+        BessScheduleConfigException = 809109
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs:446]:
```suggestion
        /// <summary>
        /// 809108 - Bess config frequency due or lapsed
        /// </summary>
        BessConfigFrequencyElapsed = 809108,
        /// <summary>
        /// 809109 - Exception occurred while processing bess config frequency
        /// </summary>
        BessConfigFrequencyProcessingException = 809109
```

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IAzureTableStorageHelper.cs [hunk 4]
```diff
@@ -1,15 +1,21 @@
...
+
         DateTime GetSinceDateTime();
 
         void SaveProductVersionDetails(List<ProductVersion> productVersions);
 
         List<ProductVersionEntities> GetLatestProductVersionDetails();
+
+        void UpsertScheduleDetail(DateTime nextSchedule, BessConfig bessConfig, bool isExecuted);
+
+        ScheduleDetailEntity GetScheduleDetail(string name);
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IAzureTableStorageHelper.cs:19]:
```suggestion
        ScheduleDetailEntity GetScheduleDetail(string configName);
```

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs [hunk 5]
```diff
@@ -82,5 +92,85 @@ private bool IsValidJson(string json, string fileName)
...
+                    
+                    if (CheckSchedule(config, existingScheduleDetail)) //Check if config schedule is missed or if it's due for the same day.
+                    {
+                        /* -- save details to msg queue --
+                         *
+                         *
+                         *
+                         */
+
+                        logger.LogInformation(EventIds.BessConfigFrequencyElapsed.ToEventId(), "Config for Name : {Name} | Frequency : {Frequency} | ScheduleTime : {ScheduleTime} | executed at Timestamp: {Timestamp} | _X-Correlation-ID : {CorrelationId}", config.Name, config.Frequency, nextOccurrence, DateTime.UtcNow, CommonHelper.CorrelationID);
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs:122]:
```suggestion
                        logger.LogInformation(EventIds.BessConfigFrequencyElapsed.ToEventId(), "Bess Config Name: {Name} with CRON ({Frequency}), Schedule At : {ScheduleTime}, Executed At : {Timestamp} | _X-Correlation-ID : {CorrelationId}", config.Name, config.Frequency, nextOccurrence, DateTime.UtcNow, CommonHelper.CorrelationID);
```

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs [hunk 6]
```diff
@@ -82,5 +92,85 @@ private bool IsValidJson(string json, string fileName)
...
+                            azureTableStorageHelper.UpsertScheduleDetail(nextOccurrence, config, false);
+                        }
+                    }
+                }
+
+                return true;
+            }
+            catch (Exception ex)
+            {
+                logger.LogError(EventIds.BessScheduleConfigException.ToEventId(), "Exception at schedule config details {DateTime} | {ErrorMessage} | _X-Correlation-ID : {CorrelationId}", DateTime.UtcNow, ex.Message, CommonHelper.CorrelationID);
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs:138]:
```suggestion
                logger.LogError(EventIds.BessScheduleConfigException.ToEventId(), "Exception occurred while processing Bess config {DateTime} | {ErrorMessage} | _X-Correlation-ID : {CorrelationId}", DateTime.UtcNow, ex.Message, CommonHelper.CorrelationID);
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs (hunk 3 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs (hunk 5 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs (hunk 6 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IAzureTableStorageHelper.cs (hunk 4 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IAzureTableStorageHelper.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs [hunk 7]
```diff
@@ -176,5 +187,200 @@ public void WhenContainerHasNoConfigs_ThenConfigIsNotAddedToList()
             dictionary.Add("Empty.json", InvalidEmptyJson);
             return dictionary;
         }
+
+        [Test]
+        public void WhenCheckConfigFrequencyAndSaveQueueDetailsSuccessful_ThenReturnsTrue()
```
Comment (shirinbt14926) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs:192]:
```suggestion
        public void WhenCheckConfigFrequencyAndSaveQueueDetailsIsSuccessful_ThenReturnsTrue()
```

Similarly for other unit test case as well

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs [hunk 8]
```diff
@@ -176,5 +187,200 @@ public void WhenContainerHasNoConfigs_ThenConfigIsNotAddedToList()
...
+            A.CallTo(() =>
+                fakeAzureTableStorageHelper.UpsertScheduleDetail(A<DateTime>.Ignored, A<BessConfig>.Ignored, A<bool>.Ignored)).MustHaveHappenedOnceOrMore();
+
+            Assert.That(result, Is.True);
+        }
+
+        private ScheduleDetailEntity GetFakeScheduleDetailsToAddInQueue()
+        {
+            var history = new ScheduleDetailEntity
+
```
Comment (shirinbt14926) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs:297]: remove extra line

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs [hunk 9]
```diff
@@ -82,5 +92,85 @@ private bool IsValidJson(string json, string fileName)
                 return false;
             }
         }
+
+        /// <summary>
+        /// check cron expression and save bespoke details to message queue
+        /// </summary>
+        /// <param name="bessConfigs"></param>
+        /// <returns></returns>
```
Comment (shirinbt14926) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs:100]: kindly update summary

**Review by shirinbt14926 — CHANGES_REQUESTED** [review]
Minor comments please

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs (hunk 7 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs (hunk 8 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs (hunk 9 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs:100]: Resolved

**Review by harshal11869 — APPROVED** [review]
Looks Good.

**Review by shirinbt14926 — APPROVED** [review]
LGTM. Thanks @MastekRohitChettri

**Review by DanOwenUKHO — APPROVED** [review]
LGTM

**Review by Nadeem14456 — APPROVED** [review]

**Review by DanOwenUKHO — APPROVED** [review]

**Pull request merged**